xclib/evaluation/xc_metrics.py

```diff
@@ -650,3 +649,139 @@ def __init__(self, true_labels, inv_propensity_scores=None,
+    for doc_indx in nb.prange(num_docs):
+        restrict_indx = min(k, num_gt[doc_indx])
+        restricted_preds[doc_indx][: restrict_indx] = pred_indices[doc_indx][:restrict_indx]
+    return restricted_preds
+
+def calc_gt_metrics(pred_indices, true_labels, k):
+    """
+    Returns MicroRecall@GT, Recall@GT
+    Arguments:
+    ----------
+    X_indices: np.ndarray
+        2D numpy array with indices sorted in descending order according to score for each row
+    true_labels: csr_matrix
+        ground truth in sparse format
+    k: int
+        max number of preds to consider
+    Returns:
+    -------
+    (R@GT, MicroR@GT)
+    """
+    num_gt = np.array([true_labels[i].nnz for i in range(true_labels.shape[0])])
+    top_gt_indices = restict_preds_for_gt_calc(pred_indices, num_gt, k)
+    
+    recall_at_gt, micro_recall_at_gt = fast_recall_with_indices(top_gt_indices, true_labels, k)
+    return recall_at_gt[-1], micro_recall_at_gt[-1]
+@nb.njit(parallel=True)
+def _fast_recall_at_k(true_labels_indices, true_labels_indptr, pred_labels_data, pred_labels_indices, pred_labels_indptr, top_k):
+    fracs = -1 * np.ones((len(true_labels_indptr) - 1, ), dtype=np.float32)
+    for i in nb.prange(len(true_labels_indptr) - 1):
+        _true_labels = true_labels_indices[true_labels_indptr[i] : true_labels_indptr[i + 1]]
+        _data = pred_labels_data[pred_labels_indptr[i] : pred_labels_indptr[i + 1]]
+        _indices = pred_labels_indices[pred_labels_indptr[i] : pred_labels_indptr[i + 1]]
+        top_inds = np.argsort(_data)[::-1][:top_k]
+        _pred_labels = _indices[top_inds]
+        if(len(_true_labels) > 0):
+            fracs[i] = len(set(_pred_labels).intersection(set(_true_labels))) / len(_true_labels)
+    return np.mean(fracs[fracs != -1])
+
+def fast_recall_k(X, true_labels, k):
+    """
+    Compute recall@k, faster than using `recall`
+    Arguments:
+    ----------
+    X: csr_matrix
+        * csr_matrix: csr_matrix with nnz at relevant places
+    true_labels: csr_matrix
+        ground truth in sparse format
+    k: int
+        compute recall at k
+    Returns:
+    -------
+    float: recall@k
+    """
+    return _fast_recall_at_k(true_labels.indices.astype(np.int64), true_labels.indptr, 
+    X.data, X.indices.astype(np.int64), X.indptr, k)
+
+@nb.njit(parallel=True)
+def _recall_with_indices(true_labels_indices, true_labels_indptr, pred_indices):
+    print(pred_indices.shape)
+    fracs = np.zeros((pred_indices.shape[0],), dtype=np.float32)
+    for i in nb.prange(len(true_labels_indptr) - 1):
+        _true_labels = true_labels_indices[true_labels_indptr[i] : true_labels_indptr[i + 1]]
+        
+        if(len(_true_labels) > 0):
+            fracs[i] = (len(set(pred_indices[i]).intersection(set(_true_labels))) / len(_true_labels))
+    
+    return np.mean(fracs)
+
+def recall_with_indices(true_labels, pred_indices, pred_data, top_k):
```

Comment:
What is this trying to do? Compute recall with given top indices and their corresponding scores?